backend/inspirehep/pidstore/models.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2019 CERN.
+#
+# inspirehep is free software; you can redistribute it and/or modify it under
+# the terms of the MIT License; see LICENSE file for more details.
+import datetime
+
+import structlog
+from invenio_db import db
+from invenio_pidstore.errors import PIDInvalidAction
+from invenio_pidstore.models import PersistentIdentifier, PIDStatus
+from sqlalchemy.exc import IntegrityError, SQLAlchemyError
+from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy_utils.models import Timestamp
+
+from inspirehep.pidstore.errors import (
+    PidRedirectionMissing,
+    WrongPidTypeRedirection,
+    WrongRedirectionPidStatus,
+)
+
+LOGGER = structlog.getLogger()
+
+
+class InspireRedirect(db.Model, Timestamp):
+    __tablename__ = "inspire_pidstore_redirect"
+    __table_args__ = (
+        db.Index(
+            "ix_inspire_pidstore_redirect_new_record", "new_pid_type", "new_pid_value"
+        ),
+    )
+
+    original_pid_type = db.Column(db.String(6), nullable=False, primary_key=True)
+    original_pid_value = db.Column(db.String(255), nullable=False, primary_key=True)
+
+    new_pid_type = db.Column(db.String(6), nullable=False)
+    new_pid_value = db.Column(db.String(255), nullable=False)
+
+    @property
+    def new_pid(self):
+        return PersistentIdentifier.get(self.new_pid_type, self.new_pid_value)
+
+    @classmethod
+    def redirect(cls, old_pid, new_pid):
+        """Redirects pid from old_pid to new_pid"""
+        if old_pid.pid_type != new_pid.pid_type:
+            raise WrongPidTypeRedirection(old_pid, new_pid)
+
+        if (
+            not old_pid.is_registered()
+            and not old_pid.is_deleted()
+            and not old_pid.is_redirected()
+        ) or not new_pid.is_registered():
+            raise WrongRedirectionPidStatus(old_pid, new_pid)
+
+        current_redirection = cls.query.filter_by(
+            original_pid_type=old_pid.pid_type, original_pid_value=old_pid.pid_value
+        ).one_or_none()
+
+        if current_redirection:
+            if old_pid.status != PIDStatus.REDIRECTED:
+                LOGGER.warning(
+                    "There is redirection for PID while status is not correct. Trying to fix.",
```

Comment:
should this ever happen? If not, I think it's ok to raise an exception here, you shouldn't try to fix stuff "silently" that is not expected to be broken.

Reply:
:+1: